File: 0746. Min Cost Climbing Stairs/0746.java
// time complexity: O(n)
// space complexity: O(n)
class Solution {
    // dynamic programming
    public int minCostClimbingStairs(int[] cost) {
        int n = cost.length;
        // dp[i] will hold the min cost to reach step i
        int[] dp = new int[n + 1];
        dp[0] = 0;  // start before step 0
        dp[1] = 0;  // or start before step 1

        for (int i = 2; i <= n; i++) {
            // either come from i-1 or i-2
            dp[i] = Math.min(dp[i - 1] + cost[i - 1],
                    dp[i - 2] + cost[i - 2]);
        }
        return dp[n];
    }
}

File: 0746. Min Cost Climbing Stairs/0746.py
# time complexity: O(n)
# space complexity: O(1)
class Solution(object):
    def minCostClimbingStairs(self, cost):
        """
        :type cost: List[int]
        :rtype: int
        """
        n = len(cost)
        prev2, prev1 = 0, 0  # costs to reach step i-2 and i-1
        
        for i in range(2, n+1):
            curr = min(prev1 + cost[i-1],
                       prev2 + cost[i-2])
            prev2, prev1 = prev1, curr
        
        return curr